### PR TITLE
Add clang-tidy CI check

### LINKED ISSUE #390: Set up C++ linter or static analysis in CI
I haven't done a C++ project before --- is there a good linter or static analyzer that we should be running on [include/cxx.h](https://github.com/dtolnay/cxx/blob/master/include/cxx.h) and [src/cxx.cc](https://github.com/dtolnay/cxx/blob/master/src/cxx.cc) in the GitHub Actions jobs?

### PR DESCRIPTION
**Pull request by iulianR**
This actually didn't properly work on my fork. I'm a bit puzzled by Github Actions.. Let's see here.

Fixes #390.

### CODE REVIEW

**Comment by iulianR** [comment]
This works for me (Fedora though):
```
$ clang-tidy src/cxx.cc -- -Iinclude 
11533 warnings generated.
/home/master/github/cxx/src/cxx.cc:14:23: warning: do not use reinterpret_cast [cppcoreguidelines-pro-type-reinterpret-cast]
  new (s) std::string(reinterpret_cast<const char *>(ptr), len);
                      ^
/home/master/github/cxx/src/cxx.cc:22:13: warning: use a trailing return type for this function [modernize-use-trailing-return-type]
const char *cxxbridge1$cxx_string$data(const std::string &s) noexcept {
~~~~~~~~~~~~^
auto                                                                  -> const char *
...
```
Not sure what's going on there:
```
Run clang-tidy-10 src/cxx.cc -- -Iinclude
Error reading configuration from /home/runner/work/cxx/cxx/src: directory doesn't exist.
Error reading configuration from /home/runner/work/cxx/cxx/src: directory doesn't exist.
Error while processing /home/runner/work/cxx/cxx/src/cxx.cc.
error: no input files [clang-diagnostic-error]
Found compiler error(s).
error: no such file or directory: '/home/runner/work/cxx/cxx/src/cxx.cc' [clang-diagnostic-error]
error: unable to handle compilation, expected exactly one compiler job in '' [clang-diagnostic-error]
Error: Process completed with exit code 1.
```

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -94,3 +94,12 @@ jobs:
       - uses: actions/checkout@v2
       - uses: dtolnay/rust-toolchain@clippy
       - run: cargo clippy --workspace --tests -- -Dclippy::all
+
+  clang-tidy:
+    name: Clang Tidy
+    runs-on: ubuntu-latest
+    steps:
+      - name: Install clang-tidy
```
Comment (roligugus) [inline .github/workflows/ci.yml:102]: You need to check out the code with
```
 - uses: actions/checkout@v2
```

File: .clang-tidy [hunk 2]
```diff
@@ -0,0 +1 @@
+Checks: clang-diagnostic-*,clang-analyzer-*,modernize-*,cppcoreguidelines-*,bugprone-*
```
Comment (roligugus) [inline .clang-tidy:1]: Seems a bit aggressive to enable everything right off the bat. Would recommend to enable the ones that are working and then enable "the desired checks" as code is fixed.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -94,3 +94,12 @@ jobs:
...
+        run: sudo apt install clang-tidy-10
+      - name: Run clang-tidy
+        run: clang-tidy-10 src/cxx.cc -- -Iinclude
```
Comment (roligugus) [inline .github/workflows/ci.yml:105]: This only does source file, i.e. `cxx.cc`.

If you want to also clang-tidy the include, you'd have to either set `--header-include='.*'` or set the `HeaderFilterRegex` parameter in the `.clang-tidy` file.

Default value is `HeaderFilterRegex: ''`. You can see that if you do a `clang-tidy--dump-config`.

Alternatively, you can simply ignore headers and run clang-tidy directly on the `include/cxx.h` header file as well.

Comment (roligugus) [reply .github/workflows/ci.yml:105]: The next thing then would be the question, whether it's worth using a compilation database to get the right compilation settings (from tests or demo builds). Probably not at this stage. The current approach should be good enough for now.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (iulianR) [reply .github/workflows/ci.yml]: Right, this feels a bit silly now :sweat_smile:. I assumed you started from a checked out repository.

File: .clang-tidy (hunk 2 above)
Comment (iulianR) [reply .clang-tidy]: I should have made it more clear from the start that I don't plan to enable all of them. I wanted to see what pops out and then someone else can make a decision about what to disable and what to keep. Trailing return types is one of the first I usually disable as well.

**Comment by iulianR** [comment]
These are the warnings now.
I'd mostly keep `modernize-use-default-member-init`, `cppcoreguidelines-special-member-functions`, `modernize-use-override`, `modernize-use-equals-default`.
`WarningsAsErrors` in .clang-tidy probably needs to also be set for CI to fail.

```
4272 warnings generated.
/home/master/github/cxx/src/../include/cxx.h:60:7: warning: class 'Str' defines a default destructor, a copy constructor and a copy assignment operator but does not define a move constructor or a move assignment operator [cppcoreguidelines-special-member-functions]
class Str final {
      ^
/home/master/github/cxx/src/../include/cxx.h:86:10: warning: use default member initializer for 'len' [modernize-use-default-member-init]
  size_t len;
         ^
            {0}
/home/master/github/cxx/src/../include/cxx.h:92:7: warning: class 'Slice' defines a default destructor, a copy constructor and a copy assignment operator but does not define a move constructor or a move assignment operator [cppcoreguidelines-special-member-functions]
class Slice final {
      ^
/home/master/github/cxx/src/../include/cxx.h:115:10: warning: use default member initializer for 'len' [modernize-use-default-member-init]
  size_t len;
         ^
            {0}
/home/master/github/cxx/src/../include/cxx.h:162:7: warning: class 'Vec' defines a non-default destructor, a move constructor and a move assignment operator but does not define a copy constructor or a copy assignment operator [cppcoreguidelines-special-member-functions]
class Vec final {
      ^
/home/master/github/cxx/src/../include/cxx.h:250:7: warning: constructor does not initialize these fields: msg, len [cppcoreguidelines-pro-type-member-init]
class Error final : public std::exception {
      ^
/home/master/github/cxx/src/../include/cxx.h:254:3: warning: annotate this function with 'override' or (rarely) 'final' [modernize-use-override]
  ~Error() noexcept;
  ^
                    override
/home/master/github/cxx/src/../include/cxx.h:481:9: warning: use '= default' to define a trivial default constructor [modernize-use-equals-default]
Box<T>::Box() noexcept {}
        ^              ~~
                       = default;
/home/master/github/cxx/src/../include/cxx.h:486:1: warning: constructor does not initialize these fields: repr [cppcoreguidelines-pro-type-member-init]
template <typename T>
^
/home/master/github/cxx/src/../include/cxx.h:514:10: warning: do not use const_cast [cppcoreguidelines-pro-type-const-cast]
  return const_cast<T *>(const_cast<const Vec<T> *>(this)->data());
         ^
/home/master/github/cxx/src/../include/cxx.h:514:26: warning: do not use const_cast [cppcoreguidelines-pro-type-const-cast]
  return const_cast<T *>(const_cast<const Vec<T> *>(this)->data());
                         ^
/home/master/github/cxx/src/cxx.cc:64:1: warning: constructor does not initialize these fields: repr [cppcoreguidelines-pro-type-member-init]
String::String() noexcept { cxxbridge1$string$new(this); }
^
/home/master/github/cxx/src/cxx.cc:66:1: warning: constructor does not initialize these fields: repr [cppcoreguidelines-pro-type-member-init]
String::String(const String &other) noexcept {
^
/home/master/github/cxx/src/cxx.cc:70:1: warning: constructor does not initialize these fields: repr [cppcoreguidelines-pro-type-member-init]
String::String(String &&other) noexcept {
^
/home/master/github/cxx/src/cxx.cc:83:1: warning: constructor does not initialize these fields: repr [cppcoreguidelines-pro-type-member-init]
String::String(const std::string &s) { initString(this, s.data(), s.length()); }
^
/home/master/github/cxx/src/cxx.cc:85:1: warning: constructor does not initialize these fields: repr [cppcoreguidelines-pro-type-member-init]
String::String(const char *s) {
^
/home/master/github/cxx/src/cxx.cc:86:3: warning: do not implicitly decay an array into a pointer; consider using gsl::array_view or an explicit cast instead [cppcoreguidelines-pro-bounds-array-to-pointer-decay]
  assert(s != nullptr);
  ^
/usr/include/assert.h:92:51: note: expanded from macro 'assert'
      : __assert_fail (#expr, __FILE__, __LINE__, __ASSERT_FUNCTION))
                                                  ^
/usr/include/assert.h:126:30: note: expanded from macro '__ASSERT_FUNCTION'
#   define __ASSERT_FUNCTION    __extension__ __PRETTY_FUNCTION__
                                ^
/home/master/github/cxx/src/cxx.cc:90:1: warning: constructor does not initialize these fields: repr [cppcoreguidelines-pro-type-member-init]
String::String(const char *s, size_t len) {
^
/home/master/github/cxx/src/cxx.cc:91:3: warning: do not implicitly decay an array into a pointer; consider using gsl::array_view or an explicit cast instead [cppcoreguidelines-pro-bounds-array-to-pointer-decay]
  assert(s != nullptr || len == 0);
  ^
/usr/include/assert.h:92:51: note: expanded from macro 'assert'
      : __assert_fail (#expr, __FILE__, __LINE__, __ASSERT_FUNCTION))
                                                  ^
/usr/include/assert.h:126:30: note: expanded from macro '__ASSERT_FUNCTION'
#   define __ASSERT_FUNCTION    __extension__ __PRETTY_FUNCTION__
                                ^
/home/master/github/cxx/src/cxx.cc:147:3: warning: do not implicitly decay an array into a pointer; consider using gsl::array_view or an explicit cast instead [cppcoreguidelines-pro-bounds-array-to-pointer-decay]
  assert(s != nullptr);
  ^
/usr/include/assert.h:92:51: note: expanded from macro 'assert'
      : __assert_fail (#expr, __FILE__, __LINE__, __ASSERT_FUNCTION))
                                                  ^
/usr/include/assert.h:126:30: note: expanded from macro '__ASSERT_FUNCTION'
#   define __ASSERT_FUNCTION    __extension__ __PRETTY_FUNCTION__
                                ^
/home/master/github/cxx/src/cxx.cc:154:3: warning: do not implicitly decay an array into a pointer; consider using gsl::array_view or an explicit cast instead [cppcoreguidelines-pro-bounds-array-to-pointer-decay]
  assert(s != nullptr || len == 0);
  ^
/usr/include/assert.h:92:51: note: expanded from macro 'assert'
      : __assert_fail (#expr, __FILE__, __LINE__, __ASSERT_FUNCTION))
                                                  ^
/usr/include/assert.h:126:30: note: expanded from macro '__ASSERT_FUNCTION'
#   define __ASSERT_FUNCTION    __extension__ __PRETTY_FUNCTION__
                                ^
/home/master/github/cxx/src/cxx.cc:175:3: warning: initializing non-owner 'char *' with a newly created 'gsl::owner<>' [cppcoreguidelines-owning-memory]
  char *copy = new char[len];
  ^
/home/master/github/cxx/src/cxx.cc:245:38: warning: 8 is a magic number; consider replacing it with a named constant [cppcoreguidelines-avoid-magic-numbers]
static_assert(sizeof(std::string) <= 8 * sizeof(void *),
                                     ^
/home/master/github/cxx/src/cxx.cc:248:9: warning: function-like macro 'STD_VECTOR_OPS' used; consider a 'constexpr' template function [cppcoreguidelines-macro-usage]
#define STD_VECTOR_OPS(RUST_TYPE, CXX_TYPE)                                    \
        ^
/home/master/github/cxx/src/cxx.cc:281:9: warning: function-like macro 'RUST_VEC_EXTERNS' used; consider a 'constexpr' template function [cppcoreguidelines-macro-usage]
#define RUST_VEC_EXTERNS(RUST_TYPE, CXX_TYPE)                                  \
        ^
/home/master/github/cxx/src/cxx.cc:296:9: warning: function-like macro 'RUST_VEC_OPS' used; consider a 'constexpr' template function [cppcoreguidelines-macro-usage]
#define RUST_VEC_OPS(RUST_TYPE, CXX_TYPE)                                      \
        ^
/home/master/github/cxx/src/cxx.cc:327:9: warning: function-like macro 'FOR_EACH_NUMERIC' used; consider a 'constexpr' template function [cppcoreguidelines-macro-usage]
#define FOR_EACH_NUMERIC(MACRO)                                                \
        ^
/home/master/github/cxx/src/cxx.cc:339:9: warning: function-like macro 'FOR_EACH_STD_VECTOR' used; consider a 'constexpr' template function [cppcoreguidelines-macro-usage]
#define FOR_EACH_STD_VECTOR(MACRO)                                             \
        ^
/home/master/github/cxx/src/cxx.cc:345:9: warning: function-like macro 'FOR_EACH_RUST_VEC' used; consider a 'constexpr' template function [cppcoreguidelines-macro-usage]
#define FOR_EACH_RUST_VEC(MACRO)                                               \
        ^
/home/master/github/cxx/src/cxx.cc:357:19: warning: constructor does not initialize these fields: repr [cppcoreguidelines-pro-type-member-init]
FOR_EACH_RUST_VEC(RUST_VEC_OPS)
                  ^
Suppressed 4231 warnings (4231 in non-user code).
Use -header-filter=.* to display errors from all non-system headers. Use -system-headers to display errors from system headers as well.
```

**Review by dtolnay — COMMENTED** [review]
Thank you, this is great.

Could you do the following?

1. Add ignores for literally everything that currently triggers. That way we can land this without bundling any code change into the same PR or breaking the build.
2. Open a new issue with a markdown checklist (`- [ ] checklist item`) for every lint that was ignored so that we can work through them and decide whether to fix or keep ignored.

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -94,3 +94,13 @@ jobs:
...
+
+    clang-tidy:
+      name: Clang Tidy
+      runs-on: ubuntu-latest
+      steps:
+        - uses: actions/checkout@v2
+        - name: Install clang-tidy
+          run: sudo apt install clang-tidy-10
+        - name: Run clang-tidy
+          run: clang-tidy-10 src/cxx.cc --header-filter='cxx\.h' -- -Iinclude -xc++ -std=c++11
```
Comment (roligugus) [inline .github/workflows/ci.yml:106]: Minor nitpicks:
- don't think the `-Iinclude` is needed due to the way `cxx.cc` includes `cxx.h`. At the same time, it does not do any harm.
- why specifying C++ with `-xc++`? The `.cc` extension should be sufficient enough for clang to detect it's a C++ file or is it not?

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -94,3 +94,13 @@ jobs:
       - uses: actions/checkout@v2
       - uses: dtolnay/rust-toolchain@clippy
       - run: cargo clippy --workspace --tests -- -Dclippy::all
+
+    clang-tidy:
```
Comment (dtolnay) [inline .github/workflows/ci.yml:98]: This yaml block is misaligned so it hasn't run in CI.

Screenshot from 2020-11-23 17-09-05

**Review by dtolnay — CHANGES_REQUESTED** [review]

**Comment by dtolnay** [comment]
I landed this with several fixes on top in #501, and filed #500 to track resolving the currently triggered checks and selecting a smaller list of suppressions.

**Pull request merged**